### PR TITLE
NIP-54: Inline Image Metadata

### PR DESCRIPTION
**Pull request by jb55**
Inline Image metadata is additional information associated with urls in notes that
clients can use to provide a better experience when loading images.

### CODE REVIEW

**Comment by vitorpamplona** [comment]
I think this is a terrible hack that duplicates hashes and metadata across all kind-1s that cite the same URL. 

How many tag properties can be added? We need space for 5 to the least. 

```json
{
  "content": "More image metadata tests don’t mind me https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg ",
  "kind": 1,
  "tags": [
    [
      "imeta",
      "url https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg",
      "blurhash eVF$^OI:${M{o#*0-nNFxakD-?xVM}WEWB%iNKxvR-oetmo#R-aen$",
      "dim 3024x4032",
      "description <accessibility descriptor with spaces>"
      "decryptionKeys <secret> <nonce>"
      "license <CC, CC-atribution, etc>"
    ]
  ]
}
```

Needless to say, the encoding of `<label><space><info>` is extremely weak. A simple percent-encoded queryString would have been much better.

A quoted NIP-94 event inside a regular kind-1 is a superior engineering solution, IMHO.

**Comment by jb55** [comment]
> I think this is a terrible hack that duplicates hashes and metadata across all kind-1s that cite the same URL.

disagree.

> How many tag properties can be added? We need space for 5 at the least, but the default will have more once we add license attributes.

nip01 doesn't specify a limit but, I think some relays do something like 256 ? lots of room.

> Needless to say, the encoding of `<label><space><info>` is extremely weak. A simple percent-encoded queryString would have been much better.

```
let parts = field.split(" ", max: 1)
if parts != 2 return null
let name = parts[0]
let data = parts[1]
```

> A quoted NIP-94 event inside a regular kind-1 is a superior engineering solution, IMHO.

This is for clients that are not planning on implementing nip-94 like damus

**Comment by jb55** [comment]
damus is not implementing nip-94, this is for clients who want a simpler self-contained option.

**Comment by staab** [comment]
My suggestion is as self-contained as the new NIP, if a little more verbose, and it re-uses an existing part of the protocol, decreasing implementation surface area, and enabling new emergent behavior. More options = less simplicity.

**Comment by alexgleason** [comment]
I'm a huge fan of this idea.

In my perfect world attachments wouldn't be present in the content at all, and we'd include tags like this instead:

```json
{
  "content": "More image metadata tests don’t mind me",
  "kind": 1,
  "tags": [
    [
      "media",
      "image/jpeg",
      "https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg",
      "blurhash eVF$^OI:${M{o#*0-nNFxakD-?xVM}WEWB%iNKxvR-oetmo#R-aen$",
      "dim 3024x4032"
    ]
  ]
}
```

(mime type and url are positional and required, other arguments are named and optional)

However if we're set on including the URL in the content, I think this is a decent solution. I would just try to generalize it to other kinds of attachments such as videos, and ideally be able to handle whatever else you could throw at it like PDFs etc.

```json
{
  "content": "More image metadata tests don’t mind me https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg ",
  "kind": 1,
  "tags": [
    [
      "mmeta",
      "mime image/jpeg",
      "url https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg",
      "blurhash eVF$^OI:${M{o#*0-nNFxakD-?xVM}WEWB%iNKxvR-oetmo#R-aen$",
      "dim 3024x4032"
    ]
  ]
}
```

`mmeta` = media meta, and mime type is added so it can be used for videos etc.

**Comment by jb55** [comment]
good call on mime, will add

**Comment by jb55** [comment]
mmeta too, this makes sense

**Comment by jb55** [comment]
hmm well mmeta start to feel a lot like nip94 at that point.. maybe it's worth keeping imeta small and concise for images...

**Comment by staab** [comment]
> How about skipping tags if you want it contained?

I already have enough trouble parsing URLs, I don't want to do more of that.

**Comment by jb55** [comment]
> How about skipping tags if you want it contained?
> 
> ```json
> {
>   "content": "More image metadata tests don’t mind me https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg#blurhash=eVF$^OI:${M{o#*0-nNFxakD-?xVM}WEWB%iNKxvR-oetmo#R-aen$&dim=3024x4032 ",
>   "kind": 1,
>   "tags": []
> }
> ```
> 
> The URL got the meta data after a `#` which should not get sent to the server anyway but should be interpreted by the client.

Interesting idea! They would make the url look pretty bad for clients that don't support this. The tag approach seemed friendlier. plus it could get really ugly once we started adding more and more metadata fields 🤔

**Comment by Giszmo** [comment]
@jb55 @staab Clients that don't support this do not really care. Current day Snort handling https://snort.social/e/note10a7rpdvms04pdzp3k3dsvp0nqs7edzzuujs5v2phgawzpfchk6gs3f92nl without problem. Note that I did url-encode stuff, apparently correctly.

Screenshot from 2023-04-27 14-42-08

**Comment by jb55** [comment]
> @jb55 @staab Clients that don't support this do not really care. Current day Snort handling 

I was just thinking of text clients and things.

**Comment by Giszmo** [comment]
Image handling in text clients wasn't really a consideration of mine but wouldn't they ellipse shorten urls anyway? And the more technical would show the tags, so ... I'm not sure what is gained by moving this to tags.

**Comment by jb55** [comment]
definitely an interesting approach, I would have to change my code and see if its simpler. it likely is 🤔, then you wouldn't have to duplicate the url in tags.

**Comment by jb55** [comment]
ok I'm leaning toward switching to this approach. I really like it. great idea @Giszmo

**Comment by Giszmo** [comment]
@jb55 glad I could help. A mention as co-author would be appreciated.

**Comment by vitorpamplona** [comment]
@Giszmo's idea is at least easier for users to copy and paste with all the attached metadata so the user doesn't need to retype accessibility descriptors every time and the app doesn't need to compute hashes again. 

The current one requires custom copy paste code to move all tags of one URLs into a new kind 1.

**Comment by jb55** [comment]
ah crap there is one issue for me here. The tags approach allows me to start building blurhashes immediately when I receive the note over the wire (a slow operation so I need to queue it immediately). I don't do any content parsing when I receive notes since that could be potentially slow for *every* note I receive over the wire, most won't be seen. ugh!

**Comment by mikedilger** [comment]
**I like @Giszmo 's approach**: I was thinking of an 'nimage1' bech encoded tlv thing (to avoid duplicating the URL in the tags) until I read @Giszmo 's note and realized that his suggestion tops that because it is backwards compatible to all the clients. And URL fragments are client-side with no participation from the web server, so this works very well.

URL encoding/decoding libraries are widely available and it is a well known art.

**We are already stuck with early content parsing:** As for @jb55 's performance issue: Until recently I wasn't doing content parsing immediately on incoming events either, but once embedded 'nevent' links became a thing, I started doing that content parsing on all incoming events (just once, cached) so that I could start fetching these referenced events immediately.  So now that I am doing content parsing anyways, I don't mind this data being in the content.

**Comment by jb55** [comment]
@mikedilger yeah if I were to change this I would have to move content parsing to all events... I would have to test how this affects battery and performance on mobile devices which are more resource constrained than something like gossip :(

**Comment by jb55** [comment]
> We are already stuck with early content parsing

I don't think this is necessarily true. Damus lazily parses and fetches this information when we are about to scroll into view.

**Comment by jb55** [comment]
I could try this with blurhash but I notice is was pretty slow and didn't want it popping in, but I can experiment with that.

**Comment by vitorpamplona** [comment]
Don't forget to percent-encode.

**Comment by mikedilger** [comment]
> @mikedilger yeah if I were to change this I would have to move content parsing to all events... I would have to test how this affects battery and performance on mobile devices which are more resource constrained than something like gossip :(

I hear you. @bu5hm4nn argued against nevent embedded for that very reason. We didn't even want to parse every note on the desktop and weren't even thinking about limited phone resources.

**Comment by fiatjaf** [comment]
I like the tag approach more than the giant URL approach.

**Comment by fiatjaf** [comment]
Is it a dumb idea to use an offset index to refer to where in the content the media reference is? https://github.com/nostr-protocol/nips/issues/319#issuecomment-1526811385

**Comment by v0l** [comment]
I think this is a bit unnecessary, NIP94 can already do this

**Comment by jb55** [comment]
> I think this is a bit unnecessary, NIP94 can already do this

Damus has no plans on implementing nip94

**Comment by jb55** [comment]
> I like the tag approach more than the giant URL approach.

It is definitely easier for clients who don’t parse everything up front.

**Comment by jb55** [comment]
> Is it a dumb idea to use an offset index to refer to where in the content the media reference is? [#319 (comment)](https://github.com/nostr-protocol/nips/issues/319#issuecomment-1526811385)

I started implementing it this way and I ended up changing it to a url reference since it felt cleaner and required less logic.

**Comment by staab** [comment]
> Damus has no plans on implementing nip94 for inline images. I strongly disagree that we should be using nip94 this way, hence this PR.

Can you elaborate on why you disagree so strongly? I can see why it's maybe sub-optimal in terms of event size, but it seems to accomplish the same thing this PR does (using the tag approach anyway). The example in this PR would just be re-written to:

```
{
  "content": "More image metadata tests don’t mind me https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg ",
  "kind": 1,
  "tags": [
    [
      "imeta", ''{"id":"<id>","pubkey":"<pubkey>","created_at":1293762194,"kind":1063,"tags":[["url","https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg"],["m","image/jpeg"],["x","<hash>"],["size",219387],["dim","3024x4032"],["blurhash","eVF$^OI:${M{o#*0-nNFxakD-?xVM}WEWB%iNKxvR-oetmo#R-aen$"]],"content":"Alt text for image"}''
    ]
  ]
}
```

**Comment by v0l** [comment]
> 

Im not saying that everybody SHOULD use nip94, and its fine if Damus doesnt implement it either.

**Comment by v0l** [comment]
> Because it breaks images in every nostr client.

Not true, Snort already supports NIP-94, and i expect others will too.

**Comment by staab** [comment]
What? Clients will just ignore the imeta tag, the above payload will work just fine on any client that parses content. Amethyst's implementation of NIP 94 is the problem, not the spec itself.

Also, sorry, accidentally edited your comment, putting it back now.

**Comment by jb55** [comment]
Why is @staab editing my comment ? Wtf?

**Comment by jb55** [comment]
> > Because it breaks images in every nostr client.
> 
> Not true, Snort already supports NIP-94, and i expect others will too.

Every micro app from @pablof7z would have to be rewritten. Are you serious? You can’t break image urls.

**Comment by pablof7z** [comment]
> > > Because it breaks images in every nostr client.
> > 
> > 
> > Not true, Snort already supports NIP-94, and i expect others will too.
> 
> Every micro app from @pablof7z would have to be rewritten. Are you serious? You can’t break image URLs.

Thanks, @jb55.

Breaking stuff that already works is a tax levied on smaller apps; do we *really* need to break things that work?

**Comment by jb55** [comment]
I feel like I’m taking crazy pills. If this is where we’re going with nostr (breaking core things in a backwards incompatible way) I am seriously concerned with its future. You guys are two of the biggest clients and you don’t care about breaking every other nostr client? What?

**Comment by staab** [comment]
> I feel like I’m taking crazy pills. If this is where we’re going with nostr (breaking core things in a backwards incompatible way) I am seriously concerned with its future. You guys are two of the biggest clients and you don’t care about breaking every other nostr client? What?

Please, just for clarity can you explain how adding `imeta` int he way I described above breaks anything? Asserting your position that NIP 94 necessarily is backwards incompatible isn't getting us anywhere. Yes, referencing 1063's inline creates latency and breaks UX, but I don't see how standalone 1063's or `imeta` implemented using 1063's break anything.

**Comment by jb55** [comment]
If you can’t see how replacing the majority of image uploads with nip94 and how it breaks images in all other clients in a backwards incompatible way I can’t help you.

**Comment by v0l** [comment]
> I feel like I’m taking crazy pills. If this is where we’re going with nostr (breaking core things in a backwards incompatible way) I am seriously concerned with its future. You guys are two of the biggest clients and you don’t care about breaking images in every other nostr client? What?

Yea i agree, you must be taking crazy pills.

What im saying is, THIS NIP doesn't make sense because NIP-94 already supports this. If you want these extra tags i think you should use NIP-94. NOT create 2 ways to do the same thing.

Im also NOT saying that every image should be referenced via NIP-94, Snort won't be doing this for every image, only in some scenarios.

Everything is optional, if Damus doesn't support the NIP-94 note reference then thats fine.

**Comment by jb55** [comment]
> 

nip94 requires me to pull another event , why would I do this when the data is small enough to include in the tags? NIP-94 is not a good solution to this.

**Comment by jb55** [comment]
And again, implying to use nip94 is asking me to break image urls. I really don’t understand your thinking here. How can I use nip94 without breaking urls ?

**Comment by v0l** [comment]
> > 
> 
> nip94 requires me to pull another event , why would I do this when the data is small enough to include in the tags? NIP-94 is not a good solution to this.

Its no different than reactions/zaps, you have to load other events which are related.

**Comment by pablof7z** [comment]
zaps/reactions are optional

**Comment by v0l** [comment]
> zaps/reactions are optional; not the same, images are core to most messages

It's not core, parsing HTTP links is not part of the spec

**Comment by staab** [comment]
@jb55 @pablof7z please read https://github.com/nostr-protocol/nips/pull/478#issuecomment-1527662820 again, @jb55 seems to be stuck on Amethyst's implementation of NIP-94, which creates a hard requirement for clients to dereference and fetch the 1063. However, 1063 can instead be embedded into a note's tags to solve the problem this PR is solving without additional latency, and without introducing a new standard.

**Comment by vitorpamplona** [comment]
>  @jb55 seems to be stuck on Amethyst's implementation of NIP-94, which creates a hard requirement for clients to dereference and fetch the 1063

Not anymore. We enabled options for plain URLs, NIP94s, and NIP95s yesterday when we shipped to the PlayStore (APKs required NIP 94). NIP94s and NIP95 are simple quotes in the regular Kind1 text ( nostr:nevent... )

**Comment by jb55** [comment]
There are so many misunderstandings here: 

- To build a blur hash I need data on the note immediately so i can start building the image (a slow process). I don’t even parse content at this point to even find a note reference let alone pull something from the network

- The dimension information allows me to build fixed size event views, so if the note pops into view I can have everything in place without it popping around. Preloading when scrolling helps a bit but that isn’t always the case.

I try to keep things as realtime as possible so I want certain layout information immediately instead of queueing this up and hoping the event is there so I can eventually layout the note.

please stop recommending nip94, this doesn’t solve my problem.

**Comment by v0l** [comment]
> please stop recommending nip94, this doesn’t solve my problem.

This PR doesn't solve any problems for me either, NIP-94 makes many things possible.

I would probably be fine with the tags approach if NIP-94 didn't already exist.

**Comment by vitorpamplona** [comment]
Seeing Will's needs for real-time rendering, my initial reaction was to include the NIP-94 event serialized into a Kind 1 tag.  In that way, we don't need to reinvent the wheel with custom encodings and extra content parsing.

**Comment by jb55** [comment]
Whatever, it doesn’t seem like anyone is going to use this. I will just do this as a damus optimization.

**Comment by staab** [comment]
> NIP94s and NIP95 are simple quotes in the regular Kind1 text ( nostr:nevent... )

This does break UX in the way @jb55 described. I personally think that's ok, as long as it's treated as emergent behavior coming from quotes, but it shouldn't be the primary way clients reference images. Will's solution does solve the latency/context problems.

But, NIP 94 can be used as I described above to solve @jb55 problem. A 1063 can be added to a kind 1's tags, providing all the data you need without parsing content or making a network request. Dimension information can easily be added to NIP 94's metadata tags. Do you see what I mean? It feels like we're totally talking past each other, so if you could take the time to understand what I'm suggesting I think this could be a productive conversation.

**Comment by fiatjaf** [comment]
Can we please keep this open? I think it's a good one, and there are multiple client authors interested who have commented, even if some didn't like it.

I don't see a conflict between this and NIP-94 either.

I see NIP-94 being useful for standalone "file pointers", like a "go here to download this file if you are interested", a more generic and broad approach that may have multiple use cases yet to be discovered.

While this one is useful for a basic kind:1 functionality, which are embedded images and videos. It's a much more restrict, but also much more important, since kind:1 social short notes clients are what we're doing most.

**Comment by alexgleason** [comment]
I'll definitely implement whatever we decide here on the Mostr bridge. NIP-94 is a lot of unnecessary round-trips just to add metadata about media attachments on a post, especially the more attachments you have.

**Comment by vitorpamplona** [comment]
I agree with @fiatjaf that this solves a different need. It's just not clear what the need is. 

None of the solutions here will ever replace NIP-94. And NIP-94 doesn't replace urls. 

If what Will wants is just blurhash and size to load things with an enhanced experience, so be it. That's not what NIP-94 is designed for anyway.

**Comment by jb55** [comment]
re: embedded nip94. I don't anything to do with nip94 at this point, especially with it's relation to nip95. some relays (like @Semisol and mine) already block nip94 and/or nip95. I will never serialize that into my notes. nip94 people need to seriously stop brigading other NIPs and trying to force people to implement their ideas.

**Comment by staab** [comment]
> re: embedded nip94. I don't anything to do with nip94 at this point, especially with it's relation to nip95. some relays (like @Semisol and mine) already block nip94 and/or nip95. I will never serialize that into my notes. nip94 people need to seriously stop brigading other NIPs and trying to force people to implement their ideas.

I honestly don't understand this. This is not a technical or engineering position, this is an emotional one. Like it or not, you're associated with it, because it's part of the protocol. Why not use it?

**Comment by jb55** [comment]
> > re: embedded nip94. I don't anything to do with nip94 at this point, especially with it's relation to nip95. some relays (like @Semisol and mine) already block nip94 and/or nip95. I will never serialize that into my notes. nip94 people need to seriously stop brigading other NIPs and trying to force people to implement their ideas.
> 
> I honestly don't understand this. This is not a technical or engineering position, this is an emotional one. Like it or not, you're associated with it, because it's part of the protocol. Why not use it?

because we don't want to encourage file storage on relays? even if nip94 isn't that directly, there is enough bad vibes after this inline image situation and breaking image urls in all other nostr clients. we're voting with our clients and relays. we're allowed to do that. fork off if you think it's that important.

**Comment by v0l** [comment]
If this is only for loading performance, then i think the hash frag method is best, but the NIP should be renamed to "Image Loading Optimization", and not "Image Metadata".

IMO it should not be used as a way to attach any/all image metadata.

**Comment by staab** [comment]
> fork off if you think it's that important.

I don't think it's important, I think the fact that we listen to each other's ideas, be conservative with the protocol, and not make things political is important.

**Comment by jb55** [comment]
> If this is only for loading performance, then i think the hash frag method is best, but the NIP should be renamed to "Image Loading Optimization", and not "Image Metadata".
> 
> IMO it should not be used as a way to attach any/all image metadata.

re-scoping the focus to "optimization": it probably makes more sense to have it in the tags so you don't have to parse content, otherwise it makes the optimization more difficult to process right away.

**Review by staab — APPROVED** [review]

**Comment by staab** [comment]
Closed https://github.com/nostr-protocol/nips/pull/480 based on feedback, approved this one.

**Comment by vitorpamplona** [comment]
Is the hash supposed to stay? If it's about image loading optimization, the hash should be out.

**Comment by jb55** [comment]
> Is the hash supposed to stay? If it's about image loading optimization, the hash should be out.

I don't see the harm of having it here? damus will use that in the future, other clients who implement this could as well.

**Comment by vitorpamplona** [comment]
If this is just for bluehash, can we restructure the tag to avoid the extra labeling & parsing layer? 

```json
  "tags": [
    [
      "blurhash",
      "https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg",
      "eVF$^OI:${M{o#*0-nNFxakD-?xVM}WEWB%iNKxvR-oetmo#R-aen$",
      "3024x4032"
    ]
  ]
```

File: 54.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      "imeta",
+      "url https://nostr.build/i/863321bb1ae68830ebcf9a343ca0a5e0ce2323d0a55b7fbe86f7a886cf61db8d.jpg",
+      "blurhash eVF$^OI:${M{o#*0-nNFxakD-?xVM}WEWB%iNKxvR-oetmo#R-aen$",
+      "dim 3024x4032"
+    ]
+  ]
+}
+```
+
+There can be multiple imeta tags for each image url in the content.
```
[the remark this answers is not in the document]
Comment (jb55) [reply 54.md:29]: This would break images in other clients, I don't want to encourage that.

**Comment by Giszmo** [comment]
I don't understand how you get around parsing the content. Does Damus not embed nostr:nevent... and stuff? That also influences the layout for 99% of clients so to move this data to the tags to me appears not helpful.

**Comment by vitorpamplona** [comment]
> I don't understand how you get around parsing the content. Does Damus not embed nostr:nevent... and stuff? That also influences the layout for 99% of clients so to move this data to the tags to me appears not helpful.

Other stuff does influence the layout, but a "blurhash" tag/frag will be useful to have fewer things moving around in simple URL previews. We would use and support it.

**Comment by jb55** [comment]
updated to positional

**Comment by Giszmo** [comment]
@vitorpamplona I had suggested to add blurhash and dimensions to the url further up and remain convinced that's the better approach for relays that parse tags and for clients that have to figure out the content anyway as they would not have the duplicated url there. Damus shows images below the text, so it wouldn't have to load the content to know here the blurhash goes and it might use default height for the content but most clients would not show TextNotes with a default height so they would have to parse the content either way.

That said, how about providing a preview tag containing some statistical data and a blurhash if there is media content in the post?

```
tags: ["preview", "5 lines, 34 words, 230 chars", "media 1200x800 eVF$^OI:${M{o#*0-nNFxakD-?xVM}WEWB%iNKxvR-oetmo#R-aen$"],
...
```

(This example is verbose for your readability but of course we can shorten it further.)

**Comment by vitorpamplona** [comment]
@Giszmo we wouldn't use something like that. We display url previews inline for as many urls as the author includes. Also, the information of the # of characters doesn't correlate to the layout needs until we see which characters are being used.

**Comment by jb55** [comment]
do we have an alt-text spec ?

**Comment by jb55** [comment]
ok all updated, added alt-text

**Comment by staab** [comment]
> do we have an alt-text spec ? maybe image optimization is too specific. this would be a great place for alt text

You mean something like [File Metadata](https://github.com/nostr-protocol/nips/blob/master/94.md)?

**Comment by vitorpamplona** [comment]
To be clear, I don't think hash, alt-text, and any other metadata should be in this spec. This is not a replacement for file metadata

**Comment by jb55** [comment]
I am not implementing nip94 so this makes the most sense to me

**Comment by jb55** [comment]
> > do we have an alt-text spec ? maybe image optimization is too specific. this would be a great place for alt text
> 
> You mean something like [File Metadata](https://github.com/nostr-protocol/nips/blob/master/94.md)?

do you have anything better to do?

**Comment by jb55** [comment]
Clarified the difference between this and NIP94: 3c5c49be330fcaf8c4fffafe484374c6e96fa5d6

**Comment by jb55** [comment]
renamed to "Inline Image Metadata" to clarify this even further

**Comment by jb55** [comment]
It's usually never a problem, urls are a weird edge case where almost all standardized on using image URLs because it made the most sense. It is not really a spec per say, it's just a thing that all clients do. Markdown in clients was another one of those things that has subjective rendering which is why damus eventually removed it.

**Comment by vitorpamplona** [comment]
So, it's not about Image Loading Optimization thing anymore? Looks like we are back to being a copy cat of NIP-94.

**Comment by jb55** [comment]
this is for inline image metadata where some fields can be used for loading optimization yes

**Comment by vitorpamplona** [comment]
Then it should not be positional anymore. There are too many inline image metadata options for people to agree on an ordering scheme. It will be a mess. 

Keep it simple, solve your optimization needs, and address the rest of the tags outside an optimization context.

**Comment by jb55** [comment]
this is exhausting. ya'll are wasting my time. I'm closing this and my interest in submitting new NIPs has gone to 0. if anyone else wants to pick this up then so be it. please do not implement this as it is no longer what damus implements.

**Comment by jb55** [comment]
I will be uploading the old version of the spec at https://damus.io/damus-io/dips if any other clients are interested in using this information in their clients.

**Comment by vitorpamplona** [comment]
Well, you folks know we are using NIP94 directly for all kinds of metadata. I am still open to implementing something that makes sense for regular URLs as long as it doesn't end up creating a duplicated structure.

**Comment by ChristopherKing42** [comment]
> It's usually never a problem, urls are a weird edge case

Are there any other edge cases that should be clarified, so that future NIPs don't mess with them (or everyone agrees that messing with the edge case is fine)?

**Pull request closed without merging**